lib/modules.nix

```diff
@@ -475,26 +474,54 @@ rec {
           [{ inherit (module) file; inherit value; }]
         ) configs;
 
+      # Convert an option tree decl to a submodule option decl
+      optionTreeToOption = decl:
+        if isOption decl.options
+        then decl
+        else decl // {
+            options = mkOption {
+              type = types.submoduleWith {
+                modules = [ { options = decl.options; } ];
+              };
```

Comment:
This leads to an error when a users option is declared with just `types.submodule`:

```nix
let
  lib = import ./lib;
in lib.evalModules {
  modules = [
    {
      _file = "a.nix";
      options.foo = lib.mkOption {
        type = lib.types.submodule {};
        default = {};
      };
    }
    {
      _file = "b.nix";
      options.foo.bar = lib.mkOption {
        type = lib.types.str;
      };
    }
  ];
}
```

gives

```
error: A submoduleWith option is declared multiple times with conflicting shorthandOnlyDefinesConfig values
(use '--show-trace' to show detailed location information)
```

Not sure if this can be fixed nicely or if it even makes sense.

Reply:
`submodule` is the legacy one, but yeah, not great. Seems like "don't care" needs to be representable and pick `true` as the default. `false` was probably a bad idea.

Reply:
Added `null` support, but only when merging.